src/bootstrap/compile.rs

```diff
@@ -1051,9 +1051,17 @@ impl Step for Assemble {
     }
 
     fn make_run(run: RunConfig<'_>) {
-        run.builder.ensure(Assemble {
-            target_compiler: run.builder.compiler(run.builder.top_stage + 1, run.target),
-        });
+        // NOTE: `builder.compiler` has the side effect of running `Assemble`
+        let target_compiler = Compiler { stage: run.builder.top_stage + 1, host: run.target };
+        if run.builder.top_stage >= 2 {
```

Comment:
I would have expected full_bootstrap to play a role here, I think.

It also feels a little iffy that this is only changing make_run -- builder.compiler(...) invokes Assemble via ensure, which won't run this code, right?

Reply:
Correct, that's what I meant in the description by "if you don't want to change the stage numbering, then I'll need to fix the sysroot issue eventually anyway and it doesn't make sense to merge this PR". The clippy PR uses `builder.compiler` and so won't be fixed by this change.

I am not going to put a ton more time into this PR until the clippy PR is unblocked, but I'll investigate full_bootstrap at that time, I hadn't thought about it.

Reply:
Someone commented on https://github.com/rust-lang/rust/issues/90244 this was causing issues, so I've updated the PR. It now takes into account full-bootstrap; I haven't tested that yet but expect it to work. I did test that `build --stage 2 compiler/rustc` now works:
```
$ x build --stage 2 compiler/rustc
Building rustbuild
   Compiling bootstrap v0.0.0 (/home/jnelson/rust-lang/rust2/src/bootstrap)
    Finished dev [unoptimized] target(s) in 12.80s
Building stage0 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.19s
Copying stage0 std from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Building stage0 compiler artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.32s
Copying stage0 rustc from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Assembling stage1 compiler (x86_64-unknown-linux-gnu)
Building stage1 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.20s
Copying stage1 std from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Uplifting stage1 std (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
Copying stage3 std from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Building stage1 compiler artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.31s
Copying stage1 rustc from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Uplifting stage1 rustc (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
Copying stage3 rustc from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Build completed successfully in 0:00:14
```